### ISSUE TITLE
DataFrame.diff/pct_change

### ISSUE THREAD

**Issue by joshuakoh1** [javascriptdata/danfojs#383]
There's no difference function?

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.diff.html
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pct_change.html

**Comment by joshuakoh1** [comment]
@risenW if this isn't a near term priority could you point me in the direction of how to hack together a difference operation? Which functions should I chain?

Thanks!

**Comment by risenW** [comment]
> @risenW if this isn't a near term priority could you point me in the direction of how to hack together a difference operation? Which functions should I chain?
> 
> Thanks!

The might help https://ourcodeworld.com/articles/read/557/how-to-calculate-a-percentage-change-increase-and-decrease-from-2-values-in-javascript

**Comment by joshuakoh1** [comment]
@risenW I mean how do I pass n, n-1 into the same function across the frame?

**Comment by risenW** [comment]
> @risenW I mean how do I pass n, n-1 into the same function across the frame? I couldn't find anything that accesses the previous row

You can use the `apply` function to apply custom functions to rows or columns of a DataFrame. See https://danfo.jsdata.org/api-reference/dataframe/danfo.dataframe.apply

**Comment by joshuakoh1** [comment]
@risenW I'm not sure how to access the previous row. When I apply, I only get the current row but I would like a difference of the row above.

```
keys = [A,B,C]
cols = [[1,2,3],[4,5,6],[7,8,9]]

I would like to diff B/C e.g. (5-2)/(6-3) then (8-5)/(9-6)
```

**Comment by NeonSpork** [comment]
If nobody else is actively working on implementing those two functions, I can try my hand at making the equivalent function in danfo? :) 

I may need a guiding hand writing tests however, though I'll do my best!

**Comment by risenW** [comment]
> If nobody else is actively working on implementing those two functions, I can try my hand at making the equivalent function in danfo? :)
> 
> I may need a guiding hand writing tests however, though I'll do my best!

PR is definitely welcome! You can read the contributor guide here first: https://danfo.jsdata.org/contributing-guide If you need any help, just ping me in the PR

**Comment by NeonSpork** [comment]
I started digging into pandas, and essentially followed a rabbit hole further down into numpy. So this will likely take me a while, as there is a _lot_ going on in the diff function!

It would appear that the best method of implementing the `diff` function is to emulate the numpy function which uses slice to isolate the relevant row/column and calculate the difference the [python API for tensorflow essentially took that same numpy logic too.](https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/python/ops/numpy_ops/np_math_ops.py#L916-L945)

Following that logic I can leverage the TensorflowJS method `tf.slice()` to achieve something similar. I've cloned and branched off, and will let you all know of any speedbumps or hiccups! Let me know if you have any thoughts or advice @risenW / @steveoni 🚀
